net/portmapper/portmapper.go

```diff
@@ -805,33 +805,46 @@ func (c *Client) Probe(ctx context.Context) (res ProbeResult, err error) {
 			continue
 		}
 		ip = ip.Unmap()
+
+		handleUPnPResponse := func() {
+			if ip != gw {
+				// https://github.com/tailscale/tailscale/issues/5502
+				c.logf("UPnP discovery response from %v, but gateway IP is %v", ip, gw)
+			}
+			meta, err := parseUPnPDiscoResponse(buf[:n])
+			if err != nil {
+				metricUPnPParseErr.Add(1)
+				c.logf("unrecognized UPnP discovery response; ignoring: %v", err)
+				return
+			}
+			metricUPnPOK.Add(1)
+			c.logf("[v1] UPnP reply %+v, %q", meta, buf[:n])
+			res.UPnP = true
+			c.mu.Lock()
+			c.uPnPSawTime = time.Now()
+			if c.uPnPMeta != meta {
+				c.logf("UPnP meta changed: %+v", meta)
+				c.uPnPMeta = meta
+				metricUPnPUpdatedMeta.Add(1)
+			}
+			c.mu.Unlock()
+		}
+
 		port := uint16(addr.(*net.UDPAddr).Port)
 		switch port {
 		case c.upnpPort():
 			metricUPnPResponse.Add(1)
 			if mem.Contains(mem.B(buf[:n]), mem.S(":InternetGatewayDevice:")) {
-				if ip != gw {
-					// https://github.com/tailscale/tailscale/issues/5502
-					c.logf("UPnP discovery response from %v, but gateway IP is %v", ip, gw)
-				}
-				meta, err := parseUPnPDiscoResponse(buf[:n])
-				if err != nil {
-					metricUPnPParseErr.Add(1)
-					c.logf("unrecognized UPnP discovery response; ignoring: %v", err)
-					continue
-				}
-				metricUPnPOK.Add(1)
-				c.logf("[v1] UPnP reply %+v, %q", meta, buf[:n])
-				res.UPnP = true
-				c.mu.Lock()
-				c.uPnPSawTime = time.Now()
-				if c.uPnPMeta != meta {
-					c.logf("UPnP meta changed: %+v", meta)
-					c.uPnPMeta = meta
-					metricUPnPUpdatedMeta.Add(1)
-				}
-				c.mu.Unlock()
+				handleUPnPResponse()
 			}
+
+		default:
+			// https://github.com/tailscale/tailscale/issues/7377
+			if mem.Contains(mem.B(buf[:n]), mem.S(":InternetGatewayDevice:")) {
+				c.logf("UPnP discovery response from non-UPnP port %d", port)
```

Comment:
Good call; moved `metricUPnPResponse` so that it is incremented in both cases (since both are valid responses), and then added a new clientmetric to track the number of alternate port responses.